### PR TITLE
softmax migration to tch

### PR DESCRIPTION
**Pull request by AmineDiro**
solves #191  for `softmax`.

In this PR:
- Installed `tch-rs` and updated contribution guide for how to install
- Set `tch` test under the `testing` feature. I can add another flag if that's not appropriate
- Started with migrating Softmax op as an example. 

Binary ops tests are really compact because of string formatting magic. But I think that they will get a little bit more verbose 💭

### CODE REVIEW

File: crates/ratchet-core/src/tensor.rs [hunk 1]
```diff
@@ -838,6 +839,17 @@ impl Tensor {
         ))
     }
 
+    #[cfg(feature = "testing")]
+    pub fn to_tch<T: TensorDType + tch::kind::Element>(&self) -> anyhow::Result<tch::Tensor> {
```
Comment (FL33TW00D) [inline crates/ratchet-core/src/tensor.rs:843]: Should we take `self` and consume here instead of `deep_clone`?

Comment (AmineDiro) [reply crates/ratchet-core/src/tensor.rs:843]: We usually pass a `&Tensor` in the tests to get the ground truth so I don't think we can consume `self` here. Although we *may* not need the  `deep_clone` as we call already call `into_owned` which has the signature : 
```rust
    pub fn into_owned(self) -> Array<A, D>
    where
        A: Clone,
        S: Data,
    {
        S::into_owned(self)
    }
``` 

So we are cloning elements either way.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -133,6 +132,26 @@ jobs:
...
+              for line in f:
+                  if 'torch==' in line:
+                      version = line.split('==')[1].strip()
+                      break
+          response = requests.get(f'https://download.pytorch.org/libtorch/cpu/libtorch-macos-arm64-{version}.zip')
+          with open('libtorch.zip', 'wb') as f:
+              f.write(response.content)
+          with zipfile.ZipFile('libtorch.zip', 'r') as zip_ref:
+              zip_ref.extractall()
+          os.environ['LIBTORCH'] = os.path.join(os.getcwd(), f'libtorch')
```
Comment (AmineDiro) [inline .github/workflows/rust.yml:152]: If the running is MACOS maybe we should add : 
`export DYLD_LIBRARY_PATH=$LIBTORCH:$DYLD_LIBRARY_PATH`. This fixed linking issues for me

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -133,6 +132,26 @@ jobs:
...
+          import zipfile
+          import os
+      
+          version = '2.3.0' # default version
+          with open('requirements.txt') as f:
+              for line in f:
+                  if 'torch==' in line:
+                      version = line.split('==')[1].strip()
+                      break
+          response = requests.get(f'https://download.pytorch.org/libtorch/cpu/libtorch-macos-arm64-{version}.zip')
```
Comment (AmineDiro) [inline .github/workflows/rust.yml:147]: I think an arm64 version of 2.3.0 torch exists as a Python wheel but if you need to download the zip. 
This one worked for me for (x86 at least):  
`wget https://download.pytorch.org/libtorch/cpu/libtorch-shared-with-deps-2.3.0%2Bcpu.zip`

Comment (FL33TW00D) [reply .github/workflows/rust.yml:147]: @AmineDiro Still being a pain 😡

Comment (AmineDiro) [reply .github/workflows/rust.yml:147]: 😔 i cant see the runner's error. Anything of help ?

Comment (FL33TW00D) [reply .github/workflows/rust.yml:147]: image

Seems tricky, I can look into it later 🫡

Comment (AmineDiro) [reply .github/workflows/rust.yml:147]: Ohh I see that's the DYLIB export 👍🏼
